File: Java/Ram.java
// Ram.java
public class Ram extends Komponen implements Upgradeable {
    private int kapasitasGB;
    private String ddr;
    private int clockSpeed;
    
    public Ram() {
        super();
        this.kapasitasGB = 0;
        this.ddr = "";
        this.clockSpeed = 0;
    }
    
    public Ram(int kapasitasGB, String ddr, String merk, String nama, double harga, int clockSpeed) {
        super(merk, nama, harga, "RAM");
        this.kapasitasGB = kapasitasGB;
        this.ddr = ddr;
        this.clockSpeed = clockSpeed;
    }
    
    public int getKapasitasGB() {
        return kapasitasGB;
    }
    
    public void setKapasitasGB(int kapasitasGB) {
        this.kapasitasGB = kapasitasGB;
    }
    
    public String getDdr() {
        return ddr;
    }
    
    public void setDdr(String ddr) {
        this.ddr = ddr;
    }
    
    public int getClockSpeed() {
        return clockSpeed;
    }
    
    public void setClockSpeed(int clockSpeed) {
        this.clockSpeed = clockSpeed;
    }
    
    @Override
    public void tampilkanInfo() {
        System.out.println("RAM: " + merk + " " + nama + 
                           " (" + kapasitasGB + " GB, " + ddr + ", " + clockSpeed + " MHz)");
        System.out.println("    Harga: Rp" + harga);
    }
    
    @Override
    public boolean canUpgrade() {
        return clockSpeed < 5000;
    }
    
    @Override
    public void upgrade() {
        if (canUpgrade()) {
            clockSpeed += 400;
            harga += harga * 0.08;
            System.out.println("RAM berhasil di-upgrade menjadi " + clockSpeed + " MHz");
        } else {
            System.out.println("RAM tidak bisa di-upgrade lebih lanjut");
        }
    }
}

File: Python/Ram.py
# ram.py
from Komponen import Komponen
from Upgradeable import Upgradeable

class Ram(Komponen, Upgradeable):
    """RAM class with multiple inheritance"""
    
    def __init__(self, kapasitas_gb=0, ddr="", merk="", nama="", harga=0, clock_speed=3200, tipe_komponen="RAM"):
        super().__init__(merk, nama, harga, tipe_komponen)
        self.kapasitas_gb = kapasitas_gb
        self.ddr = ddr
        self.clock_speed = clock_speed
    
    def get_kapasitas_gb(self):
        return self.kapasitas_gb
    
    def set_kapasitas_gb(self, kapasitas_gb):
        self.kapasitas_gb = kapasitas_gb
    
    def get_ddr(self):
        return self.ddr
    
    def set_ddr(self, ddr):
        self.ddr = ddr
    
    def get_clock_speed(self):
        return self.clock_speed
    
    def set_clock_speed(self, clock_speed):
        self.clock_speed = clock_speed
    
    def tampilkan_info(self):
        print(f"RAM: {self.merk} {self.nama} ({self.kapasitas_gb} GB, {self.ddr}, {self.clock_speed} MHz)")
        print(f"    Harga: Rp{self.harga}")
    
    def can_upgrade(self):
        return self.clock_speed < 5000
    
    def upgrade(self):
        if self.can_upgrade():
            self.clock_speed += 400
            self.harga += self.harga * 0.08
            print(f"RAM berhasil di-upgrade menjadi {self.clock_speed} MHz")
        else:
            print("RAM tidak bisa di-upgrade lebih lanjut")